capstone/src/main/java/com/google/sps/data/RecommenderAlgorithm.java

```diff
@@ -11,49 +13,158 @@
 import com.google.appengine.api.datastore.Query.FilterPredicate;
 import com.google.appengine.api.datastore.Key;
 import com.google.appengine.api.datastore.Entity;
-import java.util.HashMap;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.FetchOptions;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.TreeMap;
+import java.util.Map;
 import java.lang.Iterable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.lang.Math;
  
 public class RecommenderAlgorithm {
- 
-    public static ArrayList<Room> recommendRooms(User user) {
+
+    // TODO: Will add the proportions in the next PR when I know exact percentages
+
+    public static List<Key> recommendRooms(User user, Integer amount) {
+        RecommenderAlgorithm recommender = null;
+        recommender.normalizeAllEmbeddedEntities();
+
+        Map<Key, Double> distances = recommender.getDistanceMaps(user, user.getEmbeddedTags());
+        Map sorted = sortByValues(distances);
+
+        List<Key> rooms = new ArrayList<>();
+        for (Map.Entry<Key, Double> entry : distances.entrySet()) {
+            rooms = recommender.findSimilarRooms(user, entry.getKey(), rooms);
+        }
+
+        if (rooms.size() < amount) {
+            return rooms;
+        }
+
+        return rooms.subList(0, amount);
+    }
+
+    public static List<Key> findSimilarRooms(User user, Key key, List<Key> rooms) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        HashMap<String, Integer> userTagMap = RecommenderAlgorithm.populateUserTagMap(user);
-        
-        Query otherUserQuery = new Query(User.USER_ENTITY_NAME);
-        PreparedQuery otherUserResults = datastore.prepare(otherUserQuery);
- 
-        User otherUser = null;
-        HashMap<String, Integer> otherUserTagMap = null;
-        HashMap<User, HashMap<String,Integer>> otherUserVectorMap = new HashMap<User, HashMap<String,Integer>>();
-        for(Entity otherUserEntity : otherUserResults.asIterable()) {
-            otherUser = new User(otherUserEntity);
-            otherUserTagMap = RecommenderAlgorithm.populateUserTagMap(otherUser);
-            otherUserVectorMap.put(otherUser, otherUserTagMap);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+
+        if (results.countEntities() == 0) {
+            return new ArrayList<>();
```

Comment:
Iterating over all entities will become too slow very quickly. This should use filters

Reply:
I couldn't figure out a filter that would work for this part